File: libcst/nodes/_op.py
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Tuple

from libcst._add_slots import add_slots
from libcst._base_visitor import CSTVisitor
from libcst.nodes._base import BaseLeaf, CSTNode, CSTValidationError
from libcst.nodes._internal import CodegenState, visit_required
from libcst.nodes._whitespace import BaseParenthesizableWhitespace, SimpleWhitespace


class _BaseOneTokenOp(CSTNode, ABC):
    """
    Any node that has a static value and needs to own whitespace on both sides.
    """

    whitespace_before: BaseParenthesizableWhitespace
    whitespace_after: BaseParenthesizableWhitespace

    def _visit_and_replace_children(self, visitor: CSTVisitor) -> "_BaseOneTokenOp":
        return self.__class__(
            whitespace_before=visit_required(
                "whitespace_before", self.whitespace_before, visitor
            ),
            whitespace_after=visit_required(
                "whitespace_after", self.whitespace_after, visitor
            ),
        )

    def _codegen(self, state: CodegenState) -> None:
        self.whitespace_before._codegen(state)
        state.tokens.append(self._get_token())
        self.whitespace_after._codegen(state)

    @abstractmethod
    def _get_token(self) -> str:
        ...


class _BaseTwoTokenOp(CSTNode, ABC):
    """
    This node ends up as two tokens, so we must preserve the whitespace
    in beteween them.
    """

    whitespace_before: BaseParenthesizableWhitespace
    whitespace_between: BaseParenthesizableWhitespace
    whitespace_after: BaseParenthesizableWhitespace

    def _validate(self) -> None:
        if self.whitespace_between.empty:
            raise CSTValidationError("Must have at least one space between not and in.")

    def _visit_and_replace_children(self, visitor: CSTVisitor) -> "_BaseTwoTokenOp":
        return self.__class__(
            whitespace_before=visit_required(
                "whitespace_before", self.whitespace_before, visitor
            ),
            whitespace_between=visit_required(
                "whitespace_between", self.whitespace_between, visitor
            ),
            whitespace_after=visit_required(
                "whitespace_after", self.whitespace_after, visitor
            ),
        )

    def _codegen(self, state: CodegenState) -> None:
        self.whitespace_before._codegen(state)
        state.tokens.append(self._get_tokens()[0])
        self.whitespace_between._codegen(state)
        state.tokens.append(self._get_tokens()[1])
        self.whitespace_after._codegen(state)

    @abstractmethod
    def _get_tokens(self) -> Tuple[str, str]:
        ...


class BaseUnaryOp(CSTNode, ABC):
    """
    Any node that has a static value used in a Unary expression.
    """

    whitespace_after: BaseParenthesizableWhitespace

    def _visit_and_replace_children(self, visitor: CSTVisitor) -> "BaseUnaryOp":
        return self.__class__(
            whitespace_after=visit_required(
                "whitespace_after", self.whitespace_after, visitor
            )
        )

    def _codegen(self, state: CodegenState) -> None:
        state.tokens.append(self._get_token())
        self.whitespace_after._codegen(state)

    @abstractmethod
    def _get_token(self) -> str:
        ...


class BaseBooleanOp(_BaseOneTokenOp, ABC):
    """
    Any node that has a static value used in a Binary expression. This node
    is purely for typing.
    """


class BaseBinaryOp(CSTNode, ABC):
    """
    Any node that has a static value used in a Binary expression. This node
    is purely for typing.
    """


class BaseCompOp(CSTNode, ABC):
    """
    Any node that has a static value used in a CompExpression. This node
    is purely for typing.
    """


class BaseAugOp(CSTNode, ABC):
    """
    Any node that has a static value used in an AugAssign. This node is purely
    for typing.
    """


@add_slots
@dataclass(frozen=True)
class Semicolon(_BaseOneTokenOp):
    """
    Used by SmallStatement as a separator between subsequent SmallStatements contained
    within a SimpleStatementLine or SimpleStatementSuite.
    """

    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace("")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace("")

    def _get_token(self) -> str:
        return ";"


@add_slots
@dataclass(frozen=True)
class Colon(_BaseOneTokenOp):
    """
    Used by Slice as a separator between subsequent Expressions, and in Lambda
    to separate arguments and body.
    """

    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace("")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace("")

    def _get_token(self) -> str:
        return ":"


@add_slots
@dataclass(frozen=True)
class Comma(_BaseOneTokenOp):
    """
    Syntactic trivia used as a separator between subsequent items in various parts of
    the grammar.

    Some use-cases are:
    - Import or ImportFrom
    - Function arguments
    - Tuple/list/set/dict elements
    """

    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace("")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace("")

    def _get_token(self) -> str:
        return ","


@add_slots
@dataclass(frozen=True)
class Dot(_BaseOneTokenOp):
    """
    Used by Attribute and DottedName as a separator between subsequent
    Name nodes.
    """

    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace("")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace("")

    def _get_token(self) -> str:
        return "."


@add_slots
@dataclass(frozen=True)
class ImportStar(BaseLeaf):
    """
    Used by ImportFrom to denote a star import.
    """

    def _codegen(self, state: CodegenState) -> None:
        state.tokens.append("*")


@add_slots
@dataclass(frozen=True)
class AssignEqual(_BaseOneTokenOp):
    """
    Used by AnnAssign to denote a single equal character when doing an
    assignment on top of a type annotation.
    """

    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "="


@dataclass(frozen=True)
class Plus(BaseUnaryOp):
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace("")

    def _get_token(self) -> str:
        return "+"


@dataclass(frozen=True)
class Minus(BaseUnaryOp):
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace("")

    def _get_token(self) -> str:
        return "-"


@dataclass(frozen=True)
class BitInvert(BaseUnaryOp):
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace("")

    def _get_token(self) -> str:
        return "~"


@dataclass(frozen=True)
class Not(BaseUnaryOp):
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "not"


@dataclass(frozen=True)
class And(BaseBooleanOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "and"


@dataclass(frozen=True)
class Or(BaseBooleanOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "or"


@dataclass(frozen=True)
class Add(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "+"


@dataclass(frozen=True)
class Subtract(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "-"


@dataclass(frozen=True)
class Multiply(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "*"


@dataclass(frozen=True)
class Divide(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "/"


@dataclass(frozen=True)
class FloorDivide(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "//"


@dataclass(frozen=True)
class Modulo(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "%"


@dataclass(frozen=True)
class Power(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "**"


@dataclass(frozen=True)
class LeftShift(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "<<"


@dataclass(frozen=True)
class RightShift(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return ">>"


@dataclass(frozen=True)
class BitOr(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "|"


@dataclass(frozen=True)
class BitAnd(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "&"


@dataclass(frozen=True)
class BitXor(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "^"


@dataclass(frozen=True)
class MatrixMultiply(BaseBinaryOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "@"


@dataclass(frozen=True)
class LessThan(BaseCompOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "<"


@dataclass(frozen=True)
class GreaterThan(BaseCompOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return ">"


@dataclass(frozen=True)
class Equal(BaseCompOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "=="


@dataclass(frozen=True)
class LessThanEqual(BaseCompOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "<="


@dataclass(frozen=True)
class GreaterThanEqual(BaseCompOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return ">="


@dataclass(frozen=True)
class NotEqual(BaseCompOp):
    """
    This node defines a static value for convenience, but in reality due to
    PEP 401 it can be one of two values, both of which should be a NotEqual
    CompOp.
    """

    value: str = "!="

    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _validate(self) -> None:
        if self.value not in ["!=", "<>"]:
            raise CSTValidationError("Invalid value for NotEqual node.")

    def _visit_and_replace_children(self, visitor: CSTVisitor) -> "BaseCompOp":
        return self.__class__(
            whitespace_before=visit_required(
                "whitespace_before", self.whitespace_before, visitor
            ),
            value=self.value,
            whitespace_after=visit_required(
                "whitespace_after", self.whitespace_after, visitor
            ),
        )

    def _codegen(self, state: CodegenState) -> None:
        self.whitespace_before._codegen(state)
        state.tokens.append(self.value)
        self.whitespace_after._codegen(state)


@dataclass(frozen=True)
class In(BaseCompOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "in"


@dataclass(frozen=True)
class NotIn(BaseCompOp, _BaseTwoTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_between: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_tokens(self) -> Tuple[str, str]:
        return ("not", "in")


@dataclass(frozen=True)
class Is(BaseCompOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "is"


@dataclass(frozen=True)
class IsNot(BaseCompOp, _BaseTwoTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_between: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_tokens(self) -> Tuple[str, str]:
        return ("is", "not")


@add_slots
@dataclass(frozen=True)
class AddAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "+="


@add_slots
@dataclass(frozen=True)
class SubtractAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "-="


@add_slots
@dataclass(frozen=True)
class MultiplyAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "*="


@add_slots
@dataclass(frozen=True)
class MatrixMultiplyAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "@="


@add_slots
@dataclass(frozen=True)
class DivideAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "/="


@add_slots
@dataclass(frozen=True)
class ModuloAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "%="


@add_slots
@dataclass(frozen=True)
class BitAndAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "&="


@add_slots
@dataclass(frozen=True)
class BitOrAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "|="


@add_slots
@dataclass(frozen=True)
class BitXorAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "^="


@add_slots
@dataclass(frozen=True)
class LeftShiftAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "<<="


@add_slots
@dataclass(frozen=True)
class RightShiftAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return ">>="


@add_slots
@dataclass(frozen=True)
class PowerAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "**="


@add_slots
@dataclass(frozen=True)
class FloorDivideAssign(BaseAugOp, _BaseOneTokenOp):
    whitespace_before: BaseParenthesizableWhitespace = SimpleWhitespace(" ")
    whitespace_after: BaseParenthesizableWhitespace = SimpleWhitespace(" ")

    def _get_token(self) -> str:
        return "//="

File: native/libcst_nodes/src/op.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

use super::{whitespace::ParenthesizableWhitespace, Codegen, CodegenState};

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct Semicolon<'a> {
    /// Any space that appears directly before this semicolon.
    pub whitespace_before: ParenthesizableWhitespace<'a>,
    /// Any space that appears directly after this semicolon.
    pub whitespace_after: ParenthesizableWhitespace<'a>,
}

impl<'a> Codegen<'a> for Semicolon<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        self.whitespace_before.codegen(state);
        state.add_token(";");
        self.whitespace_after.codegen(state);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Comma<'a> {
    /// Any space that appears directly before this comma.
    pub whitespace_before: ParenthesizableWhitespace<'a>,
    /// Any space that appears directly after this comma.
    pub whitespace_after: ParenthesizableWhitespace<'a>,
}

impl<'a> Codegen<'a> for Comma<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        self.whitespace_before.codegen(state);
        state.add_token(",");
        self.whitespace_after.codegen(state);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct AssignEqual<'a> {
    /// Any space that appears directly before this equal sign.
    pub whitespace_before: ParenthesizableWhitespace<'a>,
    /// Any space that appears directly after this equal sign.
    pub whitespace_after: ParenthesizableWhitespace<'a>,
}

impl<'a> Codegen<'a> for AssignEqual<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        self.whitespace_before.codegen(state);
        state.add_token("=");
        self.whitespace_after.codegen(state);
    }
}

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct Dot<'a> {
    /// Any space that appears directly before this dot.
    pub whitespace_before: ParenthesizableWhitespace<'a>,
    /// Any space that appears directly after this dot.
    pub whitespace_after: ParenthesizableWhitespace<'a>,
}

impl<'a> Codegen<'a> for Dot<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        self.whitespace_before.codegen(state);
        state.add_token(".");
        self.whitespace_after.codegen(state);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct ImportStar {}

impl<'a> Codegen<'a> for ImportStar {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        state.add_token("*");
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum UnaryOp<'a> {
    Plus(ParenthesizableWhitespace<'a>),
    Minus(ParenthesizableWhitespace<'a>),
    BitInvert(ParenthesizableWhitespace<'a>),
    Not(ParenthesizableWhitespace<'a>),
}

impl<'a> Codegen<'a> for UnaryOp<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        let (tok, whitespace_after) = match self {
            Self::Plus(ws) => ("+", ws),
            Self::Minus(ws) => ("-", ws),
            Self::BitInvert(ws) => ("~", ws),
            Self::Not(ws) => ("not", ws),
        };
        state.add_token(tok);
        whitespace_after.codegen(state);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BooleanOp<'a> {
    And {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Or {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
}

impl<'a> Codegen<'a> for BooleanOp<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        let (tok, ws_bef, ws_aft) = match self {
            Self::And {
                whitespace_after,
                whitespace_before,
            } => ("and", whitespace_before, whitespace_after),
            Self::Or {
                whitespace_after,
                whitespace_before,
            } => ("or", whitespace_before, whitespace_after),
        };
        ws_bef.codegen(state);
        state.add_token(tok);
        ws_aft.codegen(state);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BinaryOp<'a> {
    Add {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Subtract {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Multiply {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Divide {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    FloorDivide {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Modulo {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Power {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    LeftShift {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    RightShift {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    BitOr {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    BitAnd {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    BitXor {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    MatrixMultiply {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
}

impl<'a> Codegen<'a> for BinaryOp<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        let (tok, bef, aft) = match self {
            Self::Add {
                whitespace_before,
                whitespace_after,
            } => ("+", whitespace_before, whitespace_after),
            Self::Subtract {
                whitespace_before,
                whitespace_after,
            } => ("-", whitespace_before, whitespace_after),
            Self::Multiply {
                whitespace_before,
                whitespace_after,
            } => ("*", whitespace_before, whitespace_after),
            Self::Divide {
                whitespace_before,
                whitespace_after,
            } => ("/", whitespace_before, whitespace_after),
            Self::FloorDivide {
                whitespace_before,
                whitespace_after,
            } => ("//", whitespace_before, whitespace_after),
            Self::Modulo {
                whitespace_before,
                whitespace_after,
            } => ("%", whitespace_before, whitespace_after),
            Self::Power {
                whitespace_before,
                whitespace_after,
            } => ("**", whitespace_before, whitespace_after),
            Self::LeftShift {
                whitespace_before,
                whitespace_after,
            } => ("<<", whitespace_before, whitespace_after),
            Self::RightShift {
                whitespace_before,
                whitespace_after,
            } => (">>", whitespace_before, whitespace_after),
            Self::BitOr {
                whitespace_before,
                whitespace_after,
            } => ("|", whitespace_before, whitespace_after),
            Self::BitAnd {
                whitespace_before,
                whitespace_after,
            } => ("&", whitespace_before, whitespace_after),
            Self::BitXor {
                whitespace_before,
                whitespace_after,
            } => ("^", whitespace_before, whitespace_after),
            Self::MatrixMultiply {
                whitespace_before,
                whitespace_after,
            } => ("@", whitespace_before, whitespace_after),
        };
        bef.codegen(state);
        state.add_token(tok);
        aft.codegen(state);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum CompOp<'a> {
    LessThan {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    GreaterThan {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    LessThanEqual {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    GreaterThanEqual {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Equal {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    NotEqual {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    In {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    NotIn {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_between: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    Is {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
    IsNot {
        whitespace_before: ParenthesizableWhitespace<'a>,
        whitespace_between: ParenthesizableWhitespace<'a>,
        whitespace_after: ParenthesizableWhitespace<'a>,
    },
}

impl<'a> Codegen<'a> for CompOp<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        let (tok, bef, between, aft) = match self {
            Self::LessThan {
                whitespace_before,
                whitespace_after,
            } => ("<", whitespace_before, None, whitespace_after),
            Self::GreaterThan {
                whitespace_before,
                whitespace_after,
            } => (">", whitespace_before, None, whitespace_after),
            Self::LessThanEqual {
                whitespace_before,
                whitespace_after,
            } => ("<=", whitespace_before, None, whitespace_after),
            Self::GreaterThanEqual {
                whitespace_before,
                whitespace_after,
            } => (">=", whitespace_before, None, whitespace_after),
            Self::Equal {
                whitespace_before,
                whitespace_after,
            } => ("==", whitespace_before, None, whitespace_after),
            Self::NotEqual {
                whitespace_before,
                whitespace_after,
            } => ("!=", whitespace_before, None, whitespace_after),
            Self::In {
                whitespace_before,
                whitespace_after,
            } => ("in", whitespace_before, None, whitespace_after),
            Self::Is {
                whitespace_before,
                whitespace_after,
            } => ("is", whitespace_before, None, whitespace_after),
            Self::IsNot {
                whitespace_before,
                whitespace_between,
                whitespace_after,
            } => (
                "is",
                whitespace_before,
                Some(("not", whitespace_between)),
                whitespace_after,
            ),
            Self::NotIn {
                whitespace_before,
                whitespace_between,
                whitespace_after,
            } => (
                "not",
                whitespace_before,
                Some(("in", whitespace_between)),
                whitespace_after,
            ),
        };
        bef.codegen(state);
        state.add_token(tok);
        if let Some((second_tok, btw)) = between {
            btw.codegen(state);
            state.add_token(second_tok);
        }
        aft.codegen(state);
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Colon<'a> {
    pub whitespace_before: ParenthesizableWhitespace<'a>,
    pub whitespace_after: ParenthesizableWhitespace<'a>,
}

impl<'a> Codegen<'a> for Colon<'a> {
    fn codegen(&'a self, state: &mut CodegenState<'a>) {
        self.whitespace_before.codegen(state);
        state.add_token(":");
        self.whitespace_after.codegen(state);
    }
}
